src/target/riscv/riscv.c

```diff
@@ -2372,22 +2356,38 @@ int riscv_openocd_step(struct target *target, int current,
 	if (disable_triggers(target, trigger_state) != ERROR_OK)
 		return ERROR_FAIL;
 
-	int out = riscv_step_rtos_hart(target);
-	if (out != ERROR_OK) {
-		LOG_ERROR("unable to step rtos hart");
-		return out;
+	int out = ERROR_OK;
+	uint64_t current_mstatus;
+	RISCV_INFO(info);
+	if (info->isrmask_mode == RISCV_ISRMASK_STEPONLY) {
+		/* Disable Interrupts before stepping. */
+		uint64_t irq_disabled_mask = MSTATUS_MIE | MSTATUS_HIE | MSTATUS_SIE | MSTATUS_UIE;
+		out = riscv_interrupts_disable(target, irq_disabled_mask, &current_mstatus);
+		if (out != ERROR_OK) {
+			LOG_ERROR(" unable to disable interrupts");
+			goto _exit_riscv_openocd_step;
+		}
 	}
 
+	out = riscv_step_rtos_hart(target);
+	if (out != ERROR_OK)
+		LOG_ERROR("unable to step rtos hart");
+
 	register_cache_invalidate(target->reg_cache);
 
-	if (enable_triggers(target, trigger_state) != ERROR_OK)
-		return ERROR_FAIL;
+	if (info->isrmask_mode == RISCV_ISRMASK_STEPONLY)
+		out |= riscv_interrupts_restore(target, current_mstatus);
 
-	target->state = TARGET_RUNNING;
-	target_call_event_callbacks(target, TARGET_EVENT_RESUMED);
-	target->state = TARGET_HALTED;
-	target->debug_reason = DBG_REASON_SINGLESTEP;
-	target_call_event_callbacks(target, TARGET_EVENT_HALTED);
+_exit_riscv_openocd_step:
```

Comment:
Cosmetic stuff: The goto labels have function scope, so you can safely use a shorter name - e.g. `done` or `exit`. There won't be any conflicts with labels in different functions.